File: spec/cypress/integration/authentication_spec.js
// spec/cypress/integrations/authentication_spec.js
describe('User authentication spec', function () {
  before(() => {
    cy.app('clean')
  })

  it('allows user to register and login', function () {
    cy.visit('/user_session/new')
    cy.contains('Create a new account').click()

    cy.get('form').within(($form) => {
      cy.get('input[name="user[login]"]').type('cypress')
      cy.get('input[name="user[email]"]').type('cypress@test.ssl.com')
      cy.get('input[name="user[password]"]').type('Testing_ssl+1')
      cy.get('input[name="user[password_confirmation]"]').type('Testing_ssl+1')
      cy.get('input[name="tos"]').click()
      cy.root().submit()
    })

    cy.contains('SSL.com Customer Dashboard')
    cy.visit('/logout')
  })

  it('allows existing user to login and logout', function () {
    cy.get('form').within(($form) => {
      cy.get('input[name="user_session[login]"]').type('cypress')
      cy.get('input[name="user_session[password]"]').type('Testing_ssl+1')
      cy.root().submit()
    })

    cy.get('#content').contains('SSL.com Customer Dashboard')

    cy.contains('Logout').click()

    cy.contains('Customer login')
    cy.visit('/logout')
  })

  it('fails gracefully when attempting to reset password with nonexistent login', function () {
    cy.visit('/password_resets/new')

    cy.get('form').within(($form) => {
      cy.get('input[name="login"]').type('nonexistent')
    })
    cy.get('.password_resets_btn').click()
    cy.contains('No user was found with that login')
  })

  it('allows existing user to reset password using login', function () {
    cy.appFactories([
      ['create', 'user', {login: 'token'}]
    ])
    cy.visit('/password_resets/new')

    cy.get('form').within(($form) => {
      cy.get('input[name="login"]').type('token')
    })
    cy.get('.password_resets_btn').click()
    cy.contains('Customer login')
  })

  it('allows existing user to reset password using email', function () {
    cy.appFactories([
      ['create', 'user', {email: 'cartman@gmail.com'}]
    ])
    cy.visit('/password_resets/new')

    cy.get('form').within(($form) => {
      cy.get('input[name="email"]').type('cartman@gmail.com')
    })
    cy.get('.password_resets_btn').click()
    cy.contains('Customer login')
  })

  it('fails gracefully when attempting to reset a password with nonexistent email', function () {
    cy.visit('/password_resets/new')

    cy.get('form').within(($form) => {
      cy.get('input[name="email"]').type('nonexistent@ssl.com')
    })
    cy.get('.password_resets_btn').click()
    cy.contains('No user was found with that email')
  })

  it('requires Duo 2FA when logging in as super_user', function () {
    cy.appFactories([
      ['create', 'user', 'super_user', {login: 'pickles'}],
    ])

    cy.visit('/user_session/new')

    // Fill In And Submit Login Form
    cy.get('form').within(($form) => {
      cy.get('input[name="user_session[login]"]').type('pickles')
      cy.get('input[name="user_session[password]"]').type('Testing_ssl+1')
      cy.root().submit()
    })

    // Prompted for Duo 2FA
    cy.location().should((loc) => {
      expect(loc.pathname).to.eq('/user_session/duo')
    })
  })

  it('allows sysadmin to login as another user', function () {
    cy.appFactories([['create', 'user', 'sysadmin', {login: 'tyson' }]]).then((_results) => {
      cy.setCookie('skip_duo', 'true')
      cy.visit('/user_session/new')

      cy.get('form').within(($form) => {
        cy.get('input[name="user_session[login]"]').type('tyson')
        cy.get('input[name="user_session[password]"]').type('Testing_ssl+1')
        cy.root().submit()
      })

      cy.location().should((loc) => {
        expect(loc.pathname).to.contain('/team')
        expect(loc.pathname).to.contain('/account')
      })
      cy.get('#manage_certificates').click()
      cy.get('td.dropdown').eq(2).click()
      cy.contains('login as').click()
      cy.contains('pickles')
    })
  })
})

File: spec/features/authentication_spec.rb
require 'rails_helper'

RSpec.describe 'Authentications', type: :feature do
  let!(:user) { create(:user, :owner) }
  let!(:super_user) {create(:user, :super_user)}
  let!(:login_page) {LoginPage.new}
  let!(:header) {Header.new}
  let!(:registration_page) {RegistrationPage.new}
  let!(:reset_password_page) {ResetPasswordPage.new}

  before do
    User.any_instance.stubs(:authenticated_avatar_url).returns('https://github.blog/wp-content/uploads/2012/03/codercat.jpg?fit=896%2C896')
  end

  it 'logins in user who registers automatically', js: true do
    registering = attributes_for(:user, :owner)
    visit login_path
    click_on 'Create a new account'
    registration_page.login.set registering[:login]
    registration_page.email.set registering[:email]
    registration_page.password.set registering[:password]
    registration_page.password_confirmation.set registering[:password_confirmation]
    registration_page.terms_of_service.click
    registration_page.register.click
    expect(page).to have_content('SSL.com Customer Dashboard')
  end


  scenario 'allows existing user to login and logout' do
    login_page.load
    login_page.login_with(user)
    expect(page).to have_content("username: #{user.login}")
    header.logout.click
    expect(page).to have_content('Successfully logged out.')
  end

  scenario 'fails gracefully when attempting to reset password with nonexistent login' do
    reset_password_page.load
    reset_password_page.login.set 'nonexistent'
    reset_password_page.submit.click
    expect(page).to have_content 'No user was found with that login'
  end

  scenario 'allows existing user to reset password using login' do
    reset_password_page.load
    reset_password_page.login.set user.login
    reset_password_page.submit.click
    expect(page).to have_content 'Customer login'
  end

  scenario 'allows existing user to reset password using email' do
    reset_password_page.load
    reset_password_page.email.set user.email
    reset_password_page.submit.click
    expect(page).to have_content 'Customer login'
  end

  scenario 'fails gracefully when attempting to reset a password with nonexistent email' do
    reset_password_page.load
    reset_password_page.email.set 'nonexistent@ssl.com'
    reset_password_page.submit.click
    expect(page).to have_content 'No user was found with that email'
  end

  scenario 'requires Duo 2FA when logging in as super_user' do
    login_page.load
    login_page.login_with(super_user)
    expect(page).to have_content 'Duo 2-factor authentication setup'
  end

  xit 'disallows sysadmin to view "Send to SSL.com CA" page' do
    other = create(:user)
    as_user(create(:user, :sys_admin)) do
      visit certificate_order_path(ref: "co-10000")
      expect(page).to have_content("username: #{other.login}")
    end
  end

  scenario 'superuser 30 min session logout', authentication: true,  js: true do
    visit login_path
    login_page.login_with(super_user)
    Timecop.travel(Time.current + 30.minutes)
    refresh
    expect(current_url).to include('/user_session/new')
  end

  context 'when user visited cart' do
    xit 'redirect to cart after login', js: true do
      # Cart checkout
      visit show_cart_orders_path
      find('a#add_items_img').click
      first('img[alt="Buy sm bl"]').click
      find('#next_submit').click
      find('a#checkout_img').click
      # Login
      fill_in 'user_session_login', with: user.login
      fill_in 'user_session_password', with: user.password
      find('#btn_login').click

      expect(page).to have_current_path new_order_path
    end
  end
end
